feat: add volume idle duration cleanup feature (#2497)

haydnli-shopify opened the pull request:
# Summary

Adds automatic cleanup of unused volumes after a configurable idle duration. Users can now set `auto_cleanup_duration` for volumes; if a volume isn't used by any job for that period, it's deleted automatically.

Addresses [Feature]: Delete unused volumes after idle duration #2497.

# Changes

- Add `auto_cleanup_duration` to `VolumeConfiguration`
- Track last usage with `last_job_processed_at` in `VolumeModel`
- Background task (`process_idle_volumes`) runs every 60s to delete idle volumes
- Update job processing to refresh volume usage timestamps
- Database migration for new field
- Unit and integration tests for all scenarios
- Validation to prevent external volumes from using auto-cleanup (not supported)

# Usage

- Set `auto_cleanup_duration` in volume config (YAML or API)
- Volumes not used for the specified time are auto-deleted
- Volumes still attached or with `auto_cleanup_duration: off` are never deleted
- External volumes (with `volume_id`) cannot use auto-cleanup

# Testing

- All core logic and edge cases are covered by tests
- Validation tests ensure external volumes are properly handled

haydnli-shopify committed:
feat: add volume idle duration cleanup feature (#2497)

haydnli-shopify committed:
fix: avoid FOR UPDATE with outer join in process_idle_volumes (Postgres compatibility)

haydnli-shopify committed:
Fix formatting issues

haydnli-shopify committed:
Optimize idle volume cleanup implementation

haydnli-shopify committed:
Removed unnecessary changes and refactor implementations

haydnli-shopify committed:
Removed unnecessary changes and refactor implementations

dakota-shop commented on src/dstack/_internal/core/models/volumes.py:
```diff
@@ -44,6 +45,16 @@ class VolumeConfiguration(CoreModel):
         Field(description="The volume ID. Must be specified when registering external volumes"),
     ] = None
+    idle_duration: Annotated[
+        Optional[Union[str, int, bool]],
```
Why do we need `int`? If 'off' means disable, why do we need `bool`?

haydnli-shopify replied:
Yes, the boolean is no longer needed here, we could use string 'off' to overlap this
However, I would suggest keeping integer here, 
     since the integer here could be use for specifying idle duration in seconds directly, and -1 as a special value for specifying disabled

dakota-shop replied:
Why would it be useful to specify this value in seconds when we can force it to be supplied as a human-readable string?

dakota-shop commented on src/dstack/_internal/core/models/volumes.py:
```diff
@@ -44,6 +45,16 @@ class VolumeConfiguration(CoreModel):
         Optional[str],
         Field(description="The volume ID. Must be specified when registering external volumes"),
     ] = None
+    idle_duration: Annotated[
```
Should this be named `max_idle_duration`, or maybe `auto_cleanup_idle_duration`?

haydnli-shopify replied:
Yes, I will have all the name changed!

dakota-shop commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,116 @@
+        try:
+            await delete_volumes(session, project, volume_names)
+        except Exception as e:
+            logger.error("Failed to delete idle volumes for project %s: %s", project.name, str(e))
```
If you want to log the stacktrace here (looks like we do), use `logger.exception` and leave out `str(e)` (assuming the dstack logger is a Python-standard logging interface)

haydnli-shopify replied:
Sure, I have replace all the logger function with the correct ones

haydnli-shopify commented:
For this PR,
I've tested it locally and it's working correctly. I created test volumes with different auto-cleanup settings (30s, disabled, integer values) and verified the configuration parsing works. I tested the full lifecycle - volumes get created, go idle, and are automatically deleted after the specified duration. The background task handles all the edge cases we discussed.

If needed, I can upload my test configurations to show the testing approach.

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,104 @@
+                return
+            for volume in volumes:
+                await session.refresh(volume, ["project", "attachments"])
+                lockset.add(volume.id)
```
Should not issue multiple queries under `async with lock` since this lock is supposed to be released asap. Instead take lock, select ids, add them to lockset, release the lock, then load the volumes with related attributes in one query – see how other processing tasks do that.

haydnli-shopify replied:
Thanks for the response! I've updated it to take the lock and only select volume IDs, add them to the lockset and release the lock immediately, then load the full volumes with related attributes in a single query outside the lock. 
In this way our lock is released as soon as possible, and we still get all the data we need efficiently.

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,104 @@
+
+    for project, names in by_project.items():
+        try:
+            await delete_volumes(session, project, names)
```
delete_volumes does the locking and it's not reentrant so you cannot use it from a task that locks too. It also commits halfway through, so your select for update will end there. Better to write a new deleting function in process_idle_volumes.

haydnli-shopify replied:
I've fixed this by creating a new _delete_idle_volumes function that handles the deletion directly
this function simply marks volumes as deleted in the database, i think this could avoids the reentrant lockin here..

r4victor commented on src/tests/_internal/server/background/tasks/test_process_idle_volumes.py:
```diff
@@ -0,0 +1,163 @@
+        await session.commit()
+
+        with patch(
+            "dstack._internal.server.background.tasks.process_idle_volumes.delete_volumes"
```
Why mocking?

haydnli-shopify replied:
I've removed the mocking and updated the test
the previous commit was mocking delete_volumes and just checking that it was called, but I realized tht doesn't actually test if the volume gets deleted properly. 
the current test runs the background task and then verifies that the volume is actually marked as deleted in the database by checking volume.deleted and volume.deleted_at.

haydnli-shopify committed:
Fix volume auto-cleanup locking and mocking issues

haydnli-shopify committed:
feat: add merge migration for volume cleanup and secrets

haydnli-shopify committed:
fix: remove accidentally committed test files with non-existent imports

haydnli-shopify committed:
fix: add missing dialect_name parameter to get_locker call

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,120 @@
+
+        finally:
+            for volume_id in volume_ids:
+                lockset.discard(volume_id)
```
Can use lockset.difference_update

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,120 @@
+async def process_idle_volumes():
+    lock, lockset = get_locker(get_db().dialect_name).get_lockset(VolumeModel.__tablename__)
+    async with get_session_ctx() as session:
+        # Take lock, select IDs, add to lockset, release lock
```
```suggestion
```

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,120 @@
+
+
+async def _delete_idle_volumes(session: AsyncSession, volumes: List[VolumeModel]):
+    """Delete idle volumes without using the delete_volumes function to avoid locking conflicts."""
```
Not a useful, confusing docstring

r4victor commented on ~/.dstack/server/config.yml:
Accidentally committed?

r4victor reviewed (changes requested):
The PR is ok but it does not actually deletes idle volumes – just marks them as deleted in the db.

haydnli-shopify committed:
fix: address code review comments

haydnli-shopify committed:
fix: actually delete volumes from cloud providers

haydnli-shopify committed:
Remove redundant code changes

haydnli-shopify committed:
Remove accidentally committed local config file

r4victor commented:
The issue and the PR description say to support `idle_duration` for volumes, but the PR actually supports `auto_cleanup_duration`. Which one have you intended to support?

r4victor commented on src/dstack/_internal/server/background/tasks/process_idle_volumes.py:
```diff
@@ -0,0 +1,163 @@
+    volume = volume_model_to_volume(volume_model)
+
+    if volume.external:
+        return
```
Maybe disallow creating external volumes with auto_cleanup_duration if it's not supported?

r4victor reviewed (changes requested):
Created a volume:

```
type: volume
name: my-gcp-volume
backend: gcp
region: europe-west4
availability_zone: europe-west4-b
size: 100GB
auto_cleanup_duration: 1m
```

process_idle_volumes fails with exception `sqlalchemy.exc.MissingGreenlet: greenlet_spawn` – should be easy to reproduce if creating any volume.

r4victor commented on src/dstack/_internal/server/migrations/versions/d268739bd76b_merge_volume_cleanup_and_secrets_.py:
Avoid merge db migrations, rebase migrations on master

haydnli-shopify committed:
Fix MissingGreenlet error in process_idle_volumes

haydnli-shopify committed:
Add validation for external volumes with auto_cleanup_duration

haydnli-shopify committed:
Remove merge migration and rebase on master

haydnli-shopify committed:
Added merge migration to resolve multiple heads

haydnli-shopify commented:
@r4victor Thanks for your response, I have fixed the issue you mentioned, and the PR description has been updated! A force push happened in this fix because I removed the merge migration as requested, updated the migration chain to properly rebase on master, and added validation to prevent external volumes from using auto_cleanup_duration at creation time.

r4victor commented:
@haydnli-shopify, Just do a regular merge to pull changes from master to avoid force push. No need to rebase the branch on master since we'll merge with squash anyway.

r4victor committed:
Rebase migrations

r4victor committed:
Update idle_duration type to match updated parse_idle_duration

r4victor committed:
Refactor process_idle_volumes

r4victor committed:
Refactor External volumes check

r4victor committed:
Fix client backward compatibility

r4victor commented:
Made some adjustments, merging
